contrib/pkg/awsprivatelink/endpointvpc/add.go

```diff
@@ -0,0 +1,469 @@
+				default:
+					log.WithError(err).Fatal("Failed to authorize traffic from the endpoint VPC's CIDR block to the associated VPC's worker SG")
+				}
+			}
+		}
+	}
+
+	// Update HiveConfig
+	o.addEndpointVpcToHiveConfig()
+
+	return nil
+}
+
+func (o *endpointVPCAddOptions) addEndpointVpcToHiveConfig() {
+	log.Infof("Adding endpoint VPC %v to HiveConfig", o.endpointVpcId)
+
+	// Get HiveConfig
+	hiveConfig := &hivev1.HiveConfig{}
+	if err := o.dynamicClient.Get(context.Background(), types.NamespacedName{Name: "hive"}, hiveConfig); err != nil {
+		log.WithError(err).Fatal("Failed to get HiveConfig/hive")
+	}
+	if hiveConfig.Spec.AWSPrivateLink == nil {
+		log.Fatal(`AWS Private Link is not enabled in HiveConfig. Please call "hiveutil awsprivatelink enable" first.`)
+	}
+
+	// Get AZ of each endpoint subnet
+	var endpointSubnets []hivev1.AWSPrivateLinkSubnet
+	if err := o.endpointVpcClients.DescribeSubnetsPages(
+		&ec2.DescribeSubnetsInput{
+			SubnetIds: aws.StringSlice(o.endpointSubnetIds),
+		},
+		func(page *ec2.DescribeSubnetsOutput, lastPage bool) bool {
+			for _, subnet := range page.Subnets {
+				endpointSubnet := hivev1.AWSPrivateLinkSubnet{
+					SubnetID:         *subnet.SubnetId,
+					AvailabilityZone: *subnet.AvailabilityZone,
+				}
+				endpointSubnets = append(endpointSubnets, endpointSubnet)
+			}
+			return !lastPage
+		},
+	); err != nil {
+		log.WithError(err).Fatal("Failed to describe endpoint subnets")
+	}
+	// Sort endpoint subnets by AZ first and then by subnet ID
+	sort.Slice(endpointSubnets, func(i, j int) bool {
+		return endpointSubnets[i].AvailabilityZone+endpointSubnets[i].SubnetID <
+			endpointSubnets[j].AvailabilityZone+endpointSubnets[j].SubnetID
+	})
+
+	// Make sure the endpoint VPC is specified in HiveConfig.
+	// Append/update HiveConfig.spec.awsPrivateLink.endpointVpcInventory depending on the situation.
+	endpointVpcToAdd := hivev1.AWSPrivateLinkInventory{
+		AWSPrivateLinkVPC: hivev1.AWSPrivateLinkVPC{
+			VPCID:  o.endpointVpcId,
+			Region: o.endpointVpcRegion,
+		},
+		Subnets: endpointSubnets,
+	}
+	if idx, ok := awsutils.FindVpcInInventory(o.endpointVpcId, hiveConfig.Spec.AWSPrivateLink.EndpointVPCInventory); ok {
+		if reflect.DeepEqual(hiveConfig.Spec.AWSPrivateLink.EndpointVPCInventory[idx], endpointVpcToAdd) {
+			log.Warn("Endpoint VPC found in HiveConfig. HiveConfig unchanged.")
+			return
+		}
+		log.Warn("Endpoint VPC found in HiveConfig but needs update.")
+		hiveConfig.Spec.AWSPrivateLink.EndpointVPCInventory[idx] = endpointVpcToAdd
+	} else {
+		hiveConfig.Spec.AWSPrivateLink.EndpointVPCInventory = append(
+			hiveConfig.Spec.AWSPrivateLink.EndpointVPCInventory,
+			endpointVpcToAdd,
+		)
+		log.Debugf("Endpoint VPC added to HiveConfig")
+	}
+
+	if err := o.dynamicClient.Update(context.Background(), hiveConfig); err != nil {
+		log.WithError(err).Error("Failed to update HiveConfig/hive")
+	}
+}
+
+func addRouteToRouteTablesOfEndpointSubnets(
```

Comment:
At a glance, this and the below look identical except for the `Filters`. Can we DRY them?

Reply:
Yep. I combined the addRouteToRouteTables* functions into one. 

I also did the same for remove.go.